[Constraint system] Implement switch support for function builders.

DougGregor opened the pull request:
Implement support for switch statements within function builders. Cases can
perform arbitrary pattern matches, e.g.,

    tuplify(true) { c in
      "testSwitchCombined"
      switch e {
      case .a:
        "a"
      case .b(let i, _?), .b(let i, nil):
        i + 17
      }
    }

subject to the normal rules of switch statements. Cases within function
builders cannot, however, include “fallthrough” statements, because those
(like “break” and “continue”) are control flow.

The translation of performed for `switch` statements is similar to that of
`if` statements, using `buildEither(first:)` and `buildEither(second:)` on
the function builder type.

This is the bulk of switch support, tracked by rdar://problem/50426203.

DougGregor committed:
[Constraint system] Generalize saved ASTNode -> solution target mapping.

We are currently only using the “solution targets” map for statement
conditions, but we will need it for more entities soon. Start generalizing
now.

DougGregor committed:
[Constraint system] Implement switch support for function builders.

Implement support for switch statements within function builders. Cases can
perform arbitrary pattern matches, e.g.,

    tuplify(true) { c in
      "testSwitchCombined"
      switch e {
      case .a:
        "a"
      case .b(let i, _?), .b(let i, nil):
        i + 17
      }
    }

subject to the normal rules of switch statements. Cases within function
builders cannot, however, include “fallthrough” statements, because those
(like “break” and “continue”) are control flow.

The translation of performed for `switch` statements is similar to that of
`if` statements, using `buildEither(first:)` and `buildEither(second:)` on
the function builder type.

This is the bulk of switch support, tracked by rdar://problem/50426203.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

rjmccall commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -622,12 +622,104 @@ class BuilderClosureVisitor
+    // appropriate
+    llvm::TinyPtrVector<Expr *> injectedCaseExprs;
+    SmallVector<std::pair<Type, ConstraintLocator *>, 4> injectedCaseTerms;
+    for (unsigned idx : indices(capturedCaseVars)) {
```
Did you throw out my nice, pretty balanced tree folding? :)  It should be possible to do that here.

DougGregor replied:
I used it! It's beautiful!

DougGregor replied:
That's what your `buildWrappedChainPayload` does; you give it the index and the # of elements and it does the folding magic.

rjmccall replied:
I guess it might not matter as much as it used to.

rjmccall replied:
Oh!  Look at me, not remembering what it looked like. :)

rjmccall commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -622,12 +622,104 @@ class BuilderClosureVisitor
+      if (cs->generateConstraints(target, FreeTypeVariableBinding::Disallow)) {
+        hadError = true;
+        return nullptr;
+      }
```
Do you need an explicit one-way constraint here so that we don't back-propagate type information from the cases into the subject expression?

DougGregor replied:
Great point! Right now, there isn't a one-way constraint separating the subject expression from the patterns it would match against, which is a difference from the way `switch` statements are normally type checked and will create more complicated constraint systems. I should absolutely capture the result of this expression in a one-way constraint to prevent that. Thank you!

DougGregor replied:
Fixed in the third commit, with a lovely example where we were accepting something we shouldn't.

DougGregor committed:
[Constraint system] Introduce a one-way constraint for a switch subject.

The normal type checking of switch statements checks the switch subject
first, without context, then evaluates the cases. Introduce a one-way
constraint into the type checking of switch statements within function
builders to provide this same behavior. The difference can be observed
in code such as:

    enum E {
      case a
      case b(Int, String?)
    }

    enum E2 {
      case b(Int, String?)
    }

    func getSomeEnumOverloaded(_: Double) -> E { return .a }
    func getSomeEnumOverloaded(_: Int) -> E2 { return .b(0, nil) }

    func f() {
      switch getSomeEnumOverloaded(17) {
      case .a:  // error: no member named "a" in E2
        print("a")
      default:
        print("default")
      }
    }

When the subject expression `getSomeEnumOverloaded(17)` is resolved
without consider cases, it will select the second
`getSomeEnumOverloaded(_:)`, because the literal 17 prefers to be
an `Int`. The type checking of the first case would then fail because E2
does not contain a member named "a".

Prior to this change, the same expression within a function
builder would succeed in type checking, because the lack of case named
"a" within "E2" would make the second getSomeEnumOverloaded() unusable.

Making this code work by considering the cases along with the subject
expression is not unreasonable, and may be the right long term
direction for the language. However, it's a feature that should be
discussed separately, and the semantics should agree between function
builders and normal statements.

Big thanks to John McCall for noting the missing one-way constraint!

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test
